go-controller/pkg/ovn/controller/services/node_tracker.go

```diff
@@ -0,0 +1,210 @@
+// since need to apply those service's load balancers to the new node as well.
+type nodeTracker struct {
+	sync.Mutex
+
+	// nodes is the list of nodes we know about
+	// map of name -> info
+	nodes map[string]nodeInfo
+
+	// resyncFn is the function to call so that all service are resynced
+	resyncFn func()
+}
+
+type nodeInfo struct {
+	// the node's Name
+	name string
+	// The list of physical IPs the node has
+	nodeIPs []string
+	// The pod network subnet(s)
+	podSubnets []net.IPNet
+	// the name of the node's GatewayRouter, or "" of non-existent
+	gatewayRouterName string
+	// The name of the node's switch - never empty
+	switchName string
+}
+
+// returns a list of all ip blocks "assigned" to this node
+// includes node IPs, still as a mask-1 net
+func (ni *nodeInfo) nodeSubnets() []net.IPNet {
+	out := append([]net.IPNet{}, ni.podSubnets...)
+	for _, ipStr := range ni.nodeIPs {
+		ip := net.ParseIP(ipStr)
+		if ip := ip.To4(); ip != nil {
+			out = append(out, net.IPNet{
+				IP:   ip,
+				Mask: net.CIDRMask(32, 32),
+			})
+		} else {
+			out = append(out, net.IPNet{
+				IP:   ip,
+				Mask: net.CIDRMask(128, 128),
+			})
+		}
+	}
+
+	return out
+}
+
+func newNodeTracker(nodeInformer coreinformers.NodeInformer) *nodeTracker {
+	nt := &nodeTracker{
+		nodes: map[string]nodeInfo{},
+	}
+
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			node, ok := obj.(*v1.Node)
+			if !ok {
+				return
+			}
+			nt.updateNode(node)
+		},
+		UpdateFunc: func(old, new interface{}) {
+			oldObj, ok := old.(*v1.Node)
+			if !ok {
+				return
+			}
+			newObj, ok := new.(*v1.Node)
+			if !ok {
+				return
+			}
+			// Make sure object was actually changed and not pending deletion
+			if oldObj.GetResourceVersion() == newObj.GetResourceVersion() || !newObj.GetDeletionTimestamp().IsZero() {
+				return
+			}
+
+			nt.updateNode(newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			node, ok := obj.(*v1.Node)
+			if !ok {
+				tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					klog.Errorf("couldn't understand non-tombstone object")
+					return
+				}
+				node, ok = tombstone.Obj.(*v1.Node)
+				if !ok {
+					klog.Errorf("couldn't understand tombstone object")
+					return
+				}
+			}
+			nt.removeNode(node.Name)
+		},
+	})
+
+	return nt
+
+}
+
+// updateNodeInfo updates the node info cache, and syncs all services
+// if it changed.
+func (nt *nodeTracker) updateNodeInfo(nodeName, switchName, routerName string, nodeIPs []string, podSubnets []*net.IPNet) {
+	ni := nodeInfo{
+		name:              nodeName,
+		nodeIPs:           nodeIPs,
+		podSubnets:        make([]net.IPNet, 0, len(podSubnets)),
+		gatewayRouterName: routerName,
+		switchName:        switchName,
+	}
+	for i := range podSubnets {
+		ni.podSubnets = append(ni.podSubnets, *podSubnets[i]) // de-pointer
+	}
+
+	nt.Lock()
+	if existing, ok := nt.nodes[nodeName]; ok {
```

Comment:
SO I could be wrong but if `!ok` it means the nodeInfo was deleted from the cache for during the `updateNodeInfo` call by another thread, wouldn't we want to just return and print "nodeInfo deleted during update" rather than add back the "new" nodeInfo to the cache in that case?

Reply:
`!ok` just means the node wasn't in the cache; there's no risk of concurrent deletions since the map is protected by a lock.

All we want to do here is skip a full service sync if nothing critical changes.